shlr/windbg/iob_net.c

```diff
@@ -0,0 +1,470 @@
+#include  <r_crypto.h>
+#include  <r_hash.h>
+#include  <r_socket.h>
+#include  <r_util.h>
+
+#include "kd.h"
+#include "transport.h"
+
+typedef struct iobnet_t {
+	RSocket *sock;
+	bool isInitialized; // Data Channel
+
+	// Internal buffer
+	ut8 buf[4096];
+	int off;
+	int size;
+
+	// KDNET Crypto
+	ut8 _key[32];
+	ut8 _datakey[32];
+	ut8 _hmackey[32];
+	ut8 _version;
+} iobnet_t;
+
+static const char d32[] = "[\\]^_`abcd$$$$$$$efghijklmnopqrstuvwxyz{|}~";
+static const ut64 pow36[] = {1, 36, 1296, 46656, 1679616, 60466176, 2176782336,
+	78364164096, 2821109907456, 101559956668416, 3656158440062976,
+	131621703842267136, 4738381338321616896};
+
+static ut64 base32_decode(const char *str) {
+	ut64 ret = 0;
+	size_t i;
+	size_t len = strlen (str);
+	// up to 13 char for 64bit
+	if (len > 13) {
+		return 0;
+	}
+	for (i = 0; i < len; i++) {
+		if (str[i] < 48 || str[i] > 90) {
+			return 0;
+		}
+		ut8 v = d32[str[i] - 48];
+		if (v == '$') {
+			return 0;
+		}
+		v -= 91;
+		// overflow check
+		if (i == 12 && v > 3) {
+			return 0;
+		}
+		ret += v * pow36[len - i - 1];
+	}
+	return ret;
+}
+
+static ut8 *_hmacsha256(ut8 *key, int ksize, ut8 *msg, int msize) {
+	size_t i, blocksz = 64;
+	ut8 *dkey, *okey, *ikey, *hash;
+	RHash *ctx = r_hash_new (true, R_HASH_SHA256);
+	if (!ctx) {
+		return NULL;
+	}
+
+	dkey = calloc (1, blocksz);
```

Comment:
define and assign in the same line, you can save one line by removing the `ut8 *dkey, *key ,... from the begining of this function`